logos/logos-agent/app/sessions.py

```
@@ -1379,6 +1458,69 @@ async def _note_usage(self, session_id: int, lines: list[str]) -> None:
+        """Whether this request has attempts left at all.
+
+        Distinguishes "the replacement failed to be created" from "this
+        request has had its three goes": the first is worth waiting for, the
+        second would keep a settled session owing an answer forever.
+        """
+        ref = str(session.get("trigger_ref") or "")
+        if not ref:
+            return False
+        try:
+            return await db.attempts_for_trigger(ref) < _MAX_ATTEMPTS_PER_REQUEST
+        except Exception:
```

Ruff (0.16.3)

[warning] Do not catch blind exception: `Exception`

(BLE001)